core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournalLogWriter.java

```diff
@@ -149,24 +164,132 @@ public void close() throws IOException {
+      mJournalOutputStream = null;
       throw new IOException(ExceptionMessage.JOURNAL_WRITE_FAILURE
           .getMessageWithUrl(RuntimeConstants.ALLUXIO_DEBUG_DOCS_URL,
-              mJournalOutputStream.mCurrentLog, e.getMessage()), e);
+              currentLog, e.getMessage()), e);
+    }
+  }
+
+  /**
+   * Core logic of UFS journal recovery from UFS failures.
+   *
+   * If Alluxio stores its journals in UFS, then Alluxio needs to handle UFS failures.
+   * When UFS is dead, there is nothing Alluxio can do because Alluxio relies on UFS to
+   * persist journal entries. Consequently any metadata operation will block because Alluxio
+   * cannot flush their journal entries.
+   * Once UFS comes back online, Alluxio needs to perform the following operations:
+   * 1. Locate the most recent incomplete journal file, i.e. journal file that starts with
+   *    a valid sequence number X (hex), and ends with 0x7fffffffffffffff. The journal file
+   *    name encodes this information, i.e. X-0x7fffffffffffffff.
+   * 2. Sequentially scan all the incomplete journal file, and identify the last valid journal
+   *    entry that has been persisted in UFS. Once we know the last valid journal entry in UFS,
+   *    we know the first journal entry that is not valid in UFS. Suppose it is Y.
+   * 3. Rename the incomplete journal file to X-Y. Future journal writes will write to a new file
+   *    named Y-0x7fffffffffffffff.
+   * 4. Check whether there is any missing journal entry between Y and the oldest entry in
+   *    mEntriesToFlush, say Z. If Z > Y, then it means journal entries in [Y, Z) are
+   *    missing, and Alluxio cannot recover. Otherwise, for each journal entry in
+   *    mEntriesToFlush, if its sequence number is newer than or equal to Y, retry writing it to
+   *    UFS by calling the {@code UfsJournalLogWriter#write} method.
+   * @throws IOException if reading journal entries from UFS fails
+   */
+  private void maybeRecoverFromUfsFailures() throws IOException {
+    if (mJournalOutputStream != null) {
+      return;
+    }
+    UfsJournalSnapshot snapshot = UfsJournalSnapshot.getSnapshot(mJournal);
+    UfsJournalFile currentLog = snapshot.getCurrentLog(mJournal);
+    if (currentLog == null) {
+      return;
+    }
+    long startSeq = currentLog.getStart();
+    long lastPersistSeq = -1;
+    LOG.info("Trying to recover from previous UFS journal failure."
+        + " Scanning for the first corrupted journal entry.");
+    try (JournalReader reader = new UfsJournalReader(mJournal, startSeq, true)) {
+      JournalEntry entry;
+      while ((entry = reader.read()) != null) {
+        if (entry.getSequenceNumber() > lastPersistSeq) {
+          lastPersistSeq = entry.getSequenceNumber();
+        }
+      }
+    } catch (InvalidJournalEntryException e) {
+      LOG.info("Found first corrupted journal entry.");
+    } catch (IOException e) {
+      LOG.info("I/O error encountered while trying to read journal.");
```

Comment:
it's usually wrong to both log and throw, because the place that catches the exception will probably log as well

Reply:
Done.